plugin/src/main/java/org/opensearch/sql/plugin/datasource/OpenSearchDataSourceMetadataStorage.java

```diff
@@ -0,0 +1,244 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.commons.io.IOUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.ActionFuture;
+import org.opensearch.action.DocWriteResponse;
+import org.opensearch.action.admin.indices.create.CreateIndexRequest;
+import org.opensearch.action.admin.indices.create.CreateIndexResponse;
+import org.opensearch.action.index.IndexRequest;
+import org.opensearch.action.index.IndexResponse;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.QueryBuilders;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.sql.common.encryptor.Encryptor;
+import org.opensearch.sql.datasource.DataSourceMetadataStorage;
+import org.opensearch.sql.datasource.model.DataSourceMetadata;
+import org.opensearch.sql.datasource.model.auth.AuthenticationType;
+import org.opensearch.sql.plugin.utils.XContentParserUtils;
+
+public class OpenSearchDataSourceMetadataStorage implements DataSourceMetadataStorage {
+
+  public static final String DATASOURCE_INDEX_NAME = ".ql-datasources";
+  private static final String DATASOURCE_INDEX_MAPPING_FILE_NAME = "datasources-index-mapping.yml";
+  private static final String DATASOURCE_INDEX_SETTINGS_FILE_NAME
+      = "datasources-index-settings.yml";
+  private static final Logger LOG = LogManager.getLogger();
+  private final Client client;
+  private final ClusterService clusterService;
+
+  private final Encryptor encryptor;
+
+  /**
+   * This class implements DataSourceMetadataStorage interface
+   * using OpenSearch as underlying storage.
+   *
+   * @param client         opensearch NodeClient.
+   * @param clusterService ClusterService.
+   * @param encryptor      Encryptor.
+   */
+  public OpenSearchDataSourceMetadataStorage(Client client, ClusterService clusterService,
+                                             Encryptor encryptor) {
+    this.client = client;
+    this.clusterService = clusterService;
+    this.encryptor = encryptor;
+  }
+
+  @Override
+  public List<DataSourceMetadata> getDataSourceMetadata() {
+    if (!this.clusterService.state().routingTable().hasIndex(DATASOURCE_INDEX_NAME)) {
+      createDataSourcesIndex();
+      return Collections.emptyList();
+    }
+    return searchInDataSourcesIndex(QueryBuilders.matchAllQuery());
```

Comment:
this search action could be tranlate to query execution plan, and reuse our query execution layer and query optimization.

Reply:
Any advantage in doing so ? 
Also I need to stash the thread context to take actions on system index (.ql-datasources), which can complicate query execution.

Reply:
https://github.com/opensearch-project/sql/issues/1474